File: retro-math.js
// SPDX-License-Identifier: Apache-2.0
(function (global, factory) {
  const api = factory();
  if (typeof module !== "undefined" && module.exports) module.exports = api;
  global.GalaxyRetroMath = api;
})(typeof globalThis !== "undefined" ? globalThis : this, function () {
  "use strict";

  const ELITE_GALAXY1 = Object.freeze([0x5a4a, 0x0248, 0xb753]);
  const BAM_QUARTER = 0x40000000;
  const BAM_HALF = 0x80000000;
  const U32_MASK = 0xffffffffn;
  const U16_MASK = 0xffffn;
  const U64_MAX = 0xffffffffffffffffn;
  const CORDIC_K_Q30 = 652032874n;
  const CORDIC_ATAN_BAM = Object.freeze([
    536870912n, 316933406n, 167458907n, 85004756n, 42667331n, 21354465n,
    10679838n, 5340245n, 2670163n, 1335087n, 667544n, 333772n,
    166886n, 83443n, 41722n, 20861n, 10430n, 5215n, 2608n, 1304n,
    652n, 326n, 163n, 81n
  ]);

  function u32(value) { return Number(BigInt(value) & U32_MASK); }
  function rol32(value, shift) {
    const x = BigInt(u32(value)), r = BigInt(shift & 31);
    if (r === 0n) return Number(x);
    return Number(((x << r) | (x >> (32n - r))) & U32_MASK);
  }
  function hash32(value) {
    let x = u32(value);
    x = (x ^ (x >>> 16)) >>> 0;
    x = Math.imul(x, 0x7feb352d) >>> 0;
    x = (x ^ (x >>> 15)) >>> 0;
    x = Math.imul(x, 0x846ca68b) >>> 0;
    return (x ^ (x >>> 16)) >>> 0;
  }
  function eliteTwist(seed) {
    if (!Array.isArray(seed) || seed.length !== 3) throw new TypeError("Elite seed must contain three words");
    const [a, b, c] = seed.map(value => Number(BigInt(value) & U16_MASK));
    return [b, c, (a + b + c) & 0xffff];
  }
  function matrixMultiply(a, b) {
    const out = [[0n,0n,0n],[0n,0n,0n],[0n,0n,0n]];
    for (let i = 0; i < 3; i++) for (let j = 0; j < 3; j++) {
      let sum = 0n;
      for (let k = 0; k < 3; k++) sum += a[i][k] * b[k][j];
      out[i][j] = sum & U16_MASK;
    }
    return out;
  }
  function matrixPower(steps) {
    let n = BigInt(steps);
    if (n < 0n) throw new RangeError("Elite jump must be non-negative");
    let result = [[1n,0n,0n],[0n,1n,0n],[0n,0n,1n]];
    let base = [[0n,1n,0n],[0n,0n,1n],[1n,1n,1n]];
    while (n) {
      if (n & 1n) result = matrixMultiply(result, base);
      base = matrixMultiply(base, base);
      n >>= 1n;
    }
    return result;
  }
  function eliteJump(seed, steps) {
    const words = seed.map(value => BigInt(value) & U16_MASK);
    const matrix = matrixPower(steps);
    return matrix.map(row => Number(row.reduce((sum, value, i) => sum + value * words[i], 0n) & U16_MASK));
  }
  function sectorIndex(sector) {
    let index;
    if (typeof sector === "bigint") {
      index = sector;
    } else if (typeof sector === "number") {
      if (!Number.isSafeInteger(sector)) {
        throw new RangeError("Numeric sector index must be a safe integer; use bigint for full-u64 sectors");
      }
      index = BigInt(sector);
    } else {
      throw new TypeError("Sector index must be a bigint or safe integer number");
    }
    if (index < 0n || index > U64_MAX) throw new RangeError("Sector index must fit u64");
    return index;
  }
  function sectorSeed(globalSeed, sector) {
    const index = sectorIndex(sector);
    const gs = u32(globalSeed);
    const base = [
      ELITE_GALAXY1[0] ^ (gs & 0xffff),
      (ELITE_GALAXY1[1] + (gs >>> 16)) & 0xffff,
      ELITE_GALAXY1[2] ^ (rol32(gs, 7) & 0xffff)
    ];
    return eliteJump(base, index & U32_MASK);
  }
  function sectorSalt(globalSeed, sector) {
    const index = sectorIndex(sector);
    const state = sectorSeed(globalSeed, index);
    const folded = ((((state[0] << 16) | state[1]) >>> 0) ^ rol32(state[2], 11)) >>> 0;
    const lo = Number(index & U32_MASK);
    const hi = Number((index >> 32n) & U32_MASK);
    return hash32(folded ^ hash32(lo ^ 0x9e3779b9) ^ hash32(hi ^ 0x85ebca6b));
  }
  function bamAdd(angle, delta) { return u32(BigInt(u32(angle)) + BigInt(u32(delta))); }
  function bamSub(angle, delta) { return u32(BigInt(u32(angle)) - BigInt(u32(delta))); }
  function bamAdvance(angle, delta, ticks) {
    const n = BigInt(ticks);
    if (n < 0n) throw new RangeError("Tick count must be non-negative");
    return Number((BigInt(u32(angle)) + BigInt(u32(delta)) * n) & U32_MASK);
  }
  function bamRetreat(angle, delta, ticks) {
    const n = BigInt(ticks);
    if (n < 0n) throw new RangeError("Tick count must be non-negative");
    return Number((BigInt(u32(angle)) - BigInt(u32(delta)) * n) & U32_MASK);
  }
  function sinCosQ30(angle) {
    const a = BigInt(u32(angle));
    let z = a < 0x80000000n ? a : a - 0x100000000n;
    let x = CORDIC_K_Q30, y = 0n;
    if (z > 0x40000000n) { z -= 0x80000000n; x = -x; }
    else if (z < -0x40000000n) { z += 0x80000000n; x = -x; }
    for (let i = 0; i < CORDIC_ATAN_BAM.length; i++) {
      const oldX = x;
      if (z >= 0n) {
        x -= y >> BigInt(i);
        y += oldX >> BigInt(i);
        z -= CORDIC_ATAN_BAM[i];
      } else {
        x += y >> BigInt(i);
        y -= oldX >> BigInt(i);
        z += CORDIC_ATAN_BAM[i];
      }
    }
    return { cos: Number(x), sin: Number(y) };
  }
  function projectQ16(radiusQ16, angle) {
    if (!Number.isInteger(radiusQ16) || radiusQ16 < -0x80000000 || radiusQ16 > 0x7fffffff) {
      throw new RangeError("Q16.16 radius must fit signed 32-bit");
    }
    const trig = sinCosQ30(angle), radius = BigInt(radiusQ16);
    return {
      x: Number((radius * BigInt(trig.cos)) >> 30n),
      y: Number((radius * BigInt(trig.sin)) >> 30n)
    };
  }
  return Object.freeze({
    ELITE_GALAXY1, BAM_QUARTER, BAM_HALF, hash32, eliteTwist, eliteJump,
    sectorSeed, sectorSalt, bamAdd, bamSub, bamAdvance, bamRetreat, sinCosQ30, projectQ16
  });
});

File: rust/src/retro_math.rs
// SPDX-License-Identifier: Apache-2.0
//! Deterministic reference math inspired by Elite's compact galaxy seed
//! recurrence and Doom's fixed-point/BAM arithmetic. This module is a
//! validation oracle; it does not replace GALAXY's physical dynamics.

pub const ELITE_GALAXY1: [u16; 3] = [0x5a4a, 0x0248, 0xb753];
pub const BAM_QUARTER: u32 = 0x4000_0000;
pub const BAM_HALF: u32 = 0x8000_0000;

const CORDIC_K_Q30: i64 = 652_032_874;
const CORDIC_ATAN_BAM: [i64; 24] = [
    536_870_912, 316_933_406, 167_458_907, 85_004_756, 42_667_331, 21_354_465,
    10_679_838, 5_340_245, 2_670_163, 1_335_087, 667_544, 333_772,
    166_886, 83_443, 41_722, 20_861, 10_430, 5_215, 2_608, 1_304,
    652, 326, 163, 81,
];

type Matrix = [[u32; 3]; 3];

pub fn elite_twist(seed: [u16; 3]) -> [u16; 3] {
    [seed[1], seed[2], seed[0].wrapping_add(seed[1]).wrapping_add(seed[2])]
}

fn matrix_mul(a: Matrix, b: Matrix) -> Matrix {
    let mut out = [[0_u32; 3]; 3];
    for i in 0..3 {
        for j in 0..3 {
            let mut sum = 0_u64;
            for k in 0..3 {
                sum += a[i][k] as u64 * b[k][j] as u64;
            }
            out[i][j] = (sum & 0xffff) as u32;
        }
    }
    out
}

fn matrix_pow(mut steps: u64) -> Matrix {
    let mut result = [[1, 0, 0], [0, 1, 0], [0, 0, 1]];
    let mut base = [[0, 1, 0], [0, 0, 1], [1, 1, 1]];
    while steps != 0 {
        if steps & 1 != 0 {
            result = matrix_mul(result, base);
        }
        base = matrix_mul(base, base);
        steps >>= 1;
    }
    result
}

pub fn elite_jump(seed: [u16; 3], steps: u64) -> [u16; 3] {
    let matrix = matrix_pow(steps);
    let mut out = [0_u16; 3];
    for i in 0..3 {
        let sum = (0..3).map(|j| matrix[i][j] as u64 * seed[j] as u64).sum::<u64>();
        out[i] = (sum & 0xffff) as u16;
    }
    out
}

fn rol32(value: u32, shift: u32) -> u32 {
    value.rotate_left(shift)
}

/// Elite's three-word recurrence supplies the low-32 procedural grammar.
/// High sector bits are deliberately mixed separately by `sector_salt` so the
/// reference layer does not inherit the recurrence's 2^32 repetition.
pub fn sector_seed(global_seed: u32, sector: u64) -> [u16; 3] {
    let base = [
        ELITE_GALAXY1[0] ^ global_seed as u16,
        ELITE_GALAXY1[1].wrapping_add((global_seed >> 16) as u16),
        ELITE_GALAXY1[2] ^ rol32(global_seed, 7) as u16,
    ];
    elite_jump(base, sector as u32 as u64)
}

/// A u32 salt suitable for XORing into GALAXY's existing particle mixer.
/// The Elite-like state is not used as a particle identity and cannot collapse
/// the full-u64 domain: both halves of the sector index feed the avalanche hash.
pub fn sector_salt(global_seed: u32, sector: u64) -> u32 {
    let state = sector_seed(global_seed, sector);
    let folded = (((state[0] as u32) << 16) | state[1] as u32) ^ rol32(state[2] as u32, 11);
    let lo = sector as u32;
    let hi = (sector >> 32) as u32;
    crate::hash32(folded ^ crate::hash32(lo ^ 0x9e37_79b9) ^ crate::hash32(hi ^ 0x85eb_ca6b))
}

pub fn bam_add(angle: u32, delta: u32) -> u32 {
    angle.wrapping_add(delta)
}

pub fn bam_sub(angle: u32, delta: u32) -> u32 {
    angle.wrapping_sub(delta)
}

pub fn bam_advance(angle: u32, delta: u32, ticks: u64) -> u32 {
    angle.wrapping_add(delta.wrapping_mul(ticks as u32))
}

pub fn bam_retreat(angle: u32, delta: u32, ticks: u64) -> u32 {
    angle.wrapping_sub(delta.wrapping_mul(ticks as u32))
}

/// Return cosine/sine in signed Q2.30 using only integer arithmetic after the
/// input BAM32 angle has been supplied. The constants are committed and shared
/// conceptually with the JavaScript implementation, making this a deterministic
/// cross-language reference rather than a fast production trig path.
pub fn sin_cos_q30(angle: u32) -> (i64, i64) {
    let mut z = if angle < BAM_HALF {
        angle as i64
    } else {
        angle as i64 - (1_i64 << 32)
    };
    let mut x = CORDIC_K_Q30;
    let mut y = 0_i64;
    if z > BAM_QUARTER as i64 {
        z -= BAM_HALF as i64;
        x = -x;
    } else if z < -(BAM_QUARTER as i64) {
        z += BAM_HALF as i64;
        x = -x;
    }
    for (i, atan) in CORDIC_ATAN_BAM.iter().enumerate() {
        let old_x = x;
        if z >= 0 {
            x -= y >> i;
            y += old_x >> i;
            z -= *atan;
        } else {
            x += y >> i;
            y -= old_x >> i;
            z += *atan;
        }
    }
    (x, y)
}

/// Project a signed Q16.16 radius with the integer CORDIC reference.
pub fn project_q16(radius_q16: i32, angle: u32) -> (i64, i64) {
    let (cos, sin) = sin_cos_q30(angle);
    let r = radius_q16 as i64;
    ((r * cos) >> 30, (r * sin) >> 30)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn jump_matches_repeated_twists() {
        let mut state = ELITE_GALAXY1;
        for steps in 0..1024_u64 {
            assert_eq!(elite_jump(ELITE_GALAXY1, steps), state);
            state = elite_twist(state);
        }
    }

    #[test]
    fn high_sector_bits_do_not_alias_in_salt() {
        assert_eq!(sector_seed(303, 0), sector_seed(303, 1_u64 << 32));
        assert_ne!(sector_salt(303, 0), sector_salt(303, 1_u64 << 32));
    }

    #[test]
    fn bam_round_trip_is_exact() {
        let start = 0x1234_5678;
        let delta = 0x00ab_cdef;
        let ticks = 1_000_003;
        assert_eq!(bam_retreat(bam_advance(start, delta, ticks), delta, ticks), start);
    }

    #[test]
    fn cardinal_cordic_vectors_are_stable() {
        assert_eq!(sin_cos_q30(0), (1_073_741_826, -76));
        assert_eq!(sin_cos_q30(BAM_QUARTER), (-76, 1_073_741_826));
        assert_eq!(sin_cos_q30(BAM_HALF), (-1_073_741_830, 79));
    }
}
